File: binding/csharp/IP2Region.Net/Internal/CacheStrategyFactory.cs
// Copyright 2023 The Ip2Region Authors. All rights reserved.
// Use of this source code is governed by a Apache2.0-style
// license that can be found in the LICENSE file.
// @Author Alan <lzh.shap@gmail.com>
// @Date   2023/07/25

using IP2Region.Net.Internal.Abstractions;
using IP2Region.Net.XDB;

namespace IP2Region.Net.Internal;

internal class CacheStrategyFactory(string xdbPath)
{
    public AbstractCacheStrategy CreateCacheStrategy(CachePolicy cachePolicy) => cachePolicy switch
    {
        CachePolicy.Content => new ContentCacheStrategy(xdbPath),
        CachePolicy.VectorIndex => new VectorIndexCacheStrategy(xdbPath),
        _ => new FileCacheStrategy(xdbPath),
    };
}

File: binding/typescript/src/cache/cache-strategy-factory.ts
import { CachePolicy } from "../models/cache-policy";
import { AbstractCacheStrategy } from "./abstract-cache-strategy";
import { ContentCacheStrategy } from "./content-cache-strategy";
import { FileCacheStrategy } from "./file-cache-strategy";
import { VectorIndexCacheStrategy } from "./vector-index-cache-strategy";

export class CacheStrategyFactory {
    constructor(private readonly xdbPath: string) {}

    public createCacheStrategy(cachePolicy: CachePolicy): AbstractCacheStrategy {
        switch (cachePolicy) {
            case CachePolicy.Content:
                return new ContentCacheStrategy(this.xdbPath);
            case CachePolicy.VectorIndex:
                return new VectorIndexCacheStrategy(this.xdbPath);
            case CachePolicy.File:
                return new FileCacheStrategy(this.xdbPath);
            default:
                throw new Error(`Unknown cachePolicy: ${cachePolicy}`);
        }
    }
}
